internal/systemd/mock_test.go

```diff
@@ -0,0 +1,203 @@
+package systemd_test
+
+import (
+	"flag"
+	"fmt"
+	"sync"
+	"testing"
+
+	"github.com/godbus/dbus/v5"
+	"github.com/godbus/dbus/v5/introspect"
+	log "github.com/sirupsen/logrus"
+	"github.com/ubuntu/adsys/internal/consts"
+	"github.com/ubuntu/adsys/internal/testutils"
+)
+
+var s systemdBus
+
+type systemdBus struct {
+	conn *dbus.Conn
+
+	nextJobID   int
+	reloadError bool
+
+	mu sync.Mutex
+}
+
+func (s *systemdBus) StartUnit(name string, _ string) (dbus.ObjectPath, *dbus.Error) {
+	if name == "not-a-service.service" {
+		return dbus.ObjectPath("/"), dbus.NewError(
+			fmt.Sprintf("%s.NoSuchUnit", consts.SystemdDbusRegisteredName),
+			[]interface{}{"Unit not-a-service.service not found."},
+		)
+	}
+
+	return s.emitJobSignals(name), nil
+}
+
+func (s *systemdBus) StopUnit(name string, _ string) (dbus.ObjectPath, *dbus.Error) {
+	if name == "not-a-service.service" {
+		return dbus.ObjectPath("/"), dbus.NewError(
+			fmt.Sprintf("%s.NoSuchUnit", consts.SystemdDbusRegisteredName),
+			[]interface{}{"Unit not-a-service.service not found."},
+		)
+	}
+
+	return s.emitJobSignals(name), nil
+}
+
+func (s *systemdBus) EnableUnitFiles(names []string, _ bool, _ bool) (bool, [][]string, *dbus.Error) {
+	if len(names) != 1 {
+		panic("method is only expected to be called with a single name")
+	}
+
+	if name := names[0]; name == "not-a-service.service" {
+		return false, nil, dbus.NewError(
+			fmt.Sprintf("%s.NoSuchUnit", consts.SystemdDbusRegisteredName),
+			[]interface{}{"Unit not-a-service.service not found."},
+		)
+	}
+
+	return true, [][]string{{"symlink", "/from/path", "/to/path"}}, nil
+}
+
+func (s *systemdBus) DisableUnitFiles(names []string, _ bool) ([][]string, *dbus.Error) {
+	if len(names) != 1 {
+		panic("method is only expected to be called with a single name")
+	}
+
+	if name := names[0]; name == "not-a-service.service" {
+		return nil, dbus.NewError(
+			fmt.Sprintf("%s.NoSuchUnit", consts.SystemdDbusRegisteredName),
+			[]interface{}{"Unit not-a-service.service not found."},
+		)
+	}
+
+	return [][]string{{"symlink", "/from/path", "/to/path"}}, nil
+}
+
+func (s *systemdBus) Reload() *dbus.Error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.reloadError {
+		return dbus.MakeFailedError(fmt.Errorf("reload error"))
+	}
+
+	return nil
+}
+
+func (s *systemdBus) setErrorOnReload() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.reloadError = true
+}
+
+func (s *systemdBus) emitJobSignals(name string) dbus.ObjectPath {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Increment job ID
+	s.nextJobID++
+	jobPath := dbus.ObjectPath(fmt.Sprintf("%s/Job/%d", consts.SystemdDbusObjectPath, s.nextJobID))
+
+	// Emit JobNew signal
+	err := s.conn.Emit(
+		dbus.ObjectPath(consts.SystemdDbusObjectPath),
+		fmt.Sprintf("%s.JobNew", consts.SystemdDbusManagerInterface),
+		s.nextJobID, jobPath, name,
+	)
+	if err != nil {
+		log.Fatalf("failed to emit JobNew signal: %v", err)
```

Comment:
`log.Fatalf` will exit(1) right away, which will then not teardown properly the `go test` framework. You should rather `panic()` which is recovered in the `go test`.